p2p/chain.py

```diff
@@ -421,9 +429,56 @@ def __init__(self,
             await self._handle_block_bodies(peer, list(cast(Tuple[eth.BlockBody], msg)))
         elif isinstance(cmd, eth.NewBlock):
             await self._handle_new_block(peer, cast(Dict[str, Any], msg))
+        elif isinstance(cmd, eth.GetBlockHeaders):
+            self._handle_get_block_headers(peer, cast(Dict[str, Any], msg))
+        elif isinstance(cmd, eth.GetBlockBodies):
+            self._handle_get_block_bodies(peer, cast(List[Hash32], msg))
+        elif isinstance(cmd, eth.GetReceipts):
+            self._handle_get_receipts(peer, cast(List[Hash32], msg))
+        elif isinstance(cmd, eth.GetNodeData):
+            self._handle_get_node_data(peer, cast(List[Hash32], msg))
         else:
             self.logger.debug("%s msg not handled yet, need to be implemented", cmd)
 
+    def _handle_get_block_headers(self, peer: ETHPeer, msg: Dict[str, Any]) -> None:
+        n = msg['block_number_or_hash']
+        if isinstance(n, bytes):
+            header = self.chaindb.get_block_header_by_hash(cast(Hash32, n))
+            n = header.block_number
+        limit = msg['max_headers']
+        if msg['reverse']:
+            block_numbers = list(reversed(range(max(0, n - limit), n + 1)))
+        else:
+            head_number = self.chaindb.get_canonical_head().block_number
+            block_numbers = list(range(n, min(head_number + 1, n + limit)))
+        headers = []
+        for i in block_numbers:
+            headers.append(self.chaindb.get_canonical_block_header_by_number(cast(BlockNumber, i)))
```

Comment:
Can we do this as a list comprehension instead of using `append`?

```python
headers = [
    self.chaindb.get_canonical_block_header_by_number(cast(BlockNumber, i))
    for i in block_numbers
]
```